ddtrace/utils/random.py

```diff
@@ -0,0 +1,64 @@
+import threading
+
+from .. import compat
+
+
+def xorshift64s():
+    """Thread-safe generator for pseudorandom 64-bit integers.
+
+    Implements the xorshift* algorithm with a non-linear transformation
+    (multiplication) applied to the result.
+
+    This implementation uses the recommended constants from Numerical
+    Recipes Chapter 7 (Ranq1 algorithm).
+
+    According to TPV, the period is approx. 1.8 x 10^19. So it should
+    not be used by an application that makes more than 10^12 calls.
+
+    To put this into perspective: we cap the max number of traces at
+    1k/s let's be conservative and say each trace contains 100 spans.
+
+    That's 100k spans/second which would be 100k + 1 calls to this fn
+    per second.
+
+    That's 10,000,000 seconds until we hit the period. That's 115 days
+    of 100k spans/second (with no application restart) until the period
+    is reached.
+    """
+    loc = threading.local()
```

Comment:
hmm. I'm not quite sure what you mean. It's used for as long as the program is running since the function is a while True generator. 🤔 

This allows the generator to be thread-safe (to my understanding) without needing a lock.

Reply:
There's nothing that is shared across threads, everything is function local AFAICT?

Reply:
Oh, right. Yes in terms of the generator by itself. But a global instance of the generator is meant to be exported and used by the application (https://github.com/DataDog/dd-trace-py/pull/1378/files#diff-d80bb4244d3f17d3de6ad1aeeabb4cb3R61-R64) so the function locals would be shared across threads.

I think I am misusing the thread local variable though. It should probably be something like:

```python
    loc = threading.local()

    while True:
        x = getattr(loc, "x", compat.getrandbits(64) ^ 4101842887655102017)
        x ^= x >> 21
        x ^= x << 35
        # Python integers are unbounded
        x &= 0xFFFFFFFFFFFFFFFF
        x ^= x >> 4
        loc.x = x
        yield (x * 2685821657736338717) & 0xFFFFFFFFFFFFFFFF
```

Reply:
re: global generator.

Would it be better to use thread locals to store a unique instance of the generator per-thread? does that save us on anything here?

Reply:
> Would it be better to use thread locals to store a unique instance of the generator per-thread? does that save us on anything here?

👍 

That's what I was going to suggest. It's going to cost the same but it's way less confusing — and also does not introduce `threading` stuff if you ever use this code in a non-threaded env (not our case I know, but just saying).